ql/src/java/org/apache/hadoop/hive/ql/parse/type/TypeCheckProcFactory.java

```diff
@@ -985,6 +986,11 @@ protected T getXpathOrFuncExprNodeDesc(ASTNode node,
             children.set(constIdx, newChild);
           }
         }
+        if(funcText.equalsIgnoreCase("case")) {
+          ListMultimap<TypeInfo, T> expressions = getTypeInfoTListMultimap((List<T>) children);
```

Comment:
making this for CASE doesn't look right - this `getTypeInfoTListMultimap` which was extracted from the IN part below - main purpose was to make all the IN values conform to the type of the value being looked up;

in case of CASE the 1st argument is a boolean condition - forcing all the other childs in that node to be boolean doesn't make sense in all cases.